oxidation/libparsec/crates/testbed/src/types.rs

```diff
@@ -0,0 +1,206 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use super::*;
+
+#[derive(Debug, Clone, serde::Serialize)]
+#[non_exhaustive] // Force use of `new()` constructor to compute crc
+pub struct TestbedTemplate {
+    pub id: &'static str,
+    pub root_signing_key: SigningKey,
+    pub devices: Vec<TestbedDeviceData>,
+    pub users: Vec<TestbedUserData>,
+    pub device_files: Vec<TestbedDeviceFileData>,
+    // TODO: finish me !
+    // pub manifests: Vec<std::sync::Arc<Manifest>>,
+    // pub vlobs: Vec<(VlobID, Vec<u8>)>,
+    // pub blocks: Vec<(BlockID, Vec<u8>)>,
+    // pub messages: Vec<(DeviceID, DateTime, Vec<u8>, MessageContent)>,
+    // pub sequester_services: Vec<(SequesterServiceID, SequesterServiceCertificate)>,
+    pub crc: u32,
+}
+
+impl TestbedTemplate {
+    pub fn device(&self, device_id: &DeviceID) -> &TestbedDeviceData {
+        self.devices
+            .iter()
+            .find(|x| x.device_id == *device_id)
+            .unwrap()
+    }
+    pub fn user(&self, user_id: &UserID) -> &TestbedUserData {
+        self.users.iter().find(|x| x.user_id == *user_id).unwrap()
+    }
+
+    pub fn new(
+        id: &'static str,
+        root_signing_key: SigningKey,
+        devices: Vec<TestbedDeviceData>,
+        users: Vec<TestbedUserData>,
+        device_files: Vec<TestbedDeviceFileData>,
+    ) -> Self {
+        let mut hasher = crc32fast::Hasher::new();
+
+        // Remember that changing the order (or adding items) to the hasher change the result !
+        hasher.update(id.as_bytes());
+        hasher.update(root_signing_key.as_ref());
+        for device in &devices {
+            hasher.update(&device.crc.to_le_bytes());
+        }
+        for user in &users {
+            hasher.update(&user.crc.to_le_bytes());
+        }
+        for device_file in &device_files {
+            hasher.update(&device_file.crc.to_le_bytes());
+        }
+
+        Self {
+            id,
+            root_signing_key,
+            devices,
+            users,
+            device_files,
+            crc: hasher.finalize(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, serde::Serialize)]
+#[non_exhaustive] // Force use of `new()` constructor to compute crc
+pub struct TestbedDeviceData {
+    pub device_id: DeviceID,
+    pub device_label: Option<DeviceLabel>,
+    pub signing_key: SigningKey,
+    pub local_symkey: SecretKey,
+    pub certif: DeviceCertificate,
+    pub raw_certif: Vec<u8>,
+    pub raw_redacted_certif: Vec<u8>,
+    pub crc: u32,
+}
+
+impl TestbedDeviceData {
+    pub(crate) fn new(
+        device_id: &DeviceID,
+        device_label: Option<DeviceLabel>,
+        signing_key: SigningKey,
+        local_symkey: SecretKey,
+        author: CertificateSignerOwned,
+        author_signkey: &SigningKey,
+        timestamp: DateTime,
```

Comment:
Given the number of params, we may want to have a builder pattern for that in a latter time when we add some other params